app/src/main/java/com/kickstarter/libs/LakeTrackingClient.kt

```diff
@@ -0,0 +1,80 @@
+package com.kickstarter.libs
+
+import android.content.Context
+import android.os.Bundle
+import android.util.Log
+import com.kickstarter.libs.qualifiers.ApplicationContext
+import com.kickstarter.libs.utils.MapUtils
+import com.kickstarter.models.User
+import com.kickstarter.services.LakeBackgroundService
+import com.kickstarter.services.firebase.dispatchJob
+import com.kickstarter.ui.IntentKey
+import io.fabric.sdk.android.Fabric
+import org.json.JSONException
+import org.json.JSONObject
+import timber.log.Timber
+import java.util.*
+
+class LakeTrackingClient(
+        @param:ApplicationContext private val context: Context,
+        currentUser: CurrentUserType,
+        build: Build,
+        currentConfig: CurrentConfigType) : TrackingClient(context, currentUser, build, currentConfig) {
+    private var loggedInUser: User? = null
+    private var config: Config? = null
+
+    init {
+
+        // Cache the most recent logged in user for default Lake properties.
+        this.currentUser.observable().subscribe { u -> this.loggedInUser = u }
+
+        // Cache the most recent config for default Lake properties.
+        this.currentConfig.observable().subscribe { c -> this.config = c }
+    }
+
+    override fun track(eventName: String, additionalProperties: Map<String, Any>) {
+        try {
+            val trackingData = getTrackingData(eventName, combinedProperties(additionalProperties))
+            val bundle = Bundle()
+            bundle.putString(IntentKey.LAKE_EVENT_NAME, eventName)
+            bundle.putString(IntentKey.LAKE_EVENT, trackingData)
+
+            val uniqueJobName = LakeBackgroundService.BASE_JOB_NAME + System.currentTimeMillis()
+            dispatchJob(this.context, LakeBackgroundService::class.java, uniqueJobName, bundle)
+            if (this.build.isDebug) {
+                Log.d(TAG, "Queued event:$trackingData")
+            }
+        } catch (e: JSONException) {
+            if (this.build.isDebug) {
+                Timber.e("Failed to encode event: $eventName")
+            }
+            Fabric.getLogger().e(TAG, "Failed to encode event: $eventName")
+        }
+
+    }
+
+    @Throws(JSONException::class)
+    private fun getTrackingData(eventName: String, newProperties: Map<String, Any>): String {
```

Comment:
This one is also very similar (with the exception of the random string). What if we have a function on the parent that does at least part of the job? So we duplicate the less amount of code possible.

Reply:
oh no these are very different! Koala's data structure is:
```
{
    "event": "event_name",
    "properties": {}
}
```

and Lake's is:
```
{
  "partition-key": "d78cc21f-1b09-4ed8-a530-d897dcaf631a",
  "data": {
    "event": "event_name",
    "properties": {
      
    }
  }
}
```

Reply:
Hmm got it!